agent/structs/config_entry_gateways.go

```diff
@@ -687,3 +690,300 @@ func (g *GatewayService) Clone() *GatewayService {
 		ServiceKind:  g.ServiceKind,
 	}
 }
+
+// APIGatewayConfigEntry manages the configuration for an API gateway service
+// with the given name.
+type APIGatewayConfigEntry struct {
+	// Kind of the config entry. This will be set to structs.APIGateway.
+	Kind string
+
+	// Name is used to match the config entry with its associated API gateway
+	// service. This should match the name provided in the service definition.
+	Name string
+
+	// Listeners is the set of listener configuration to which an API Gateway
+	// might bind.
+	Listeners []APIGatewayListener
+	// Status is the asynchronous status which an APIGateway propagates to the user.
+	Status Status
+
+	Meta               map[string]string `json:",omitempty"`
+	acl.EnterpriseMeta `hcl:",squash" mapstructure:",squash"`
+	RaftIndex
+}
+
+func (e *APIGatewayConfigEntry) GetKind() string {
+	return APIGateway
+}
+
+func (e *APIGatewayConfigEntry) GetName() string {
+	if e == nil {
+		return ""
+	}
+	return e.Name
+}
+
+func (e *APIGatewayConfigEntry) GetMeta() map[string]string {
+	if e == nil {
+		return nil
+	}
+	return e.Meta
+}
+
+func (e *APIGatewayConfigEntry) Normalize() error {
+	for i, listener := range e.Listeners {
+		protocol := strings.ToLower(string(listener.Protocol))
+		listener.Protocol = APIGatewayListenerProtocol(protocol)
+		e.Listeners[i] = listener
+	}
+	return nil
+}
+
+func (e *APIGatewayConfigEntry) Validate() error {
+	if err := e.validateListenerNames(); err != nil {
+		return err
+	}
+	if err := e.validateMergedListeners(); err != nil {
+		return err
+	}
+
+	return e.validateListeners()
+}
+
+func (e *APIGatewayConfigEntry) validateListenerNames() error {
+	listeners := make(map[string]struct{})
+	for _, listener := range e.Listeners {
+		if _, found := listeners[listener.Name]; found {
+			return fmt.Errorf("found multiple listeners with the name %q", listener.Name)
+		}
+		listeners[listener.Name] = struct{}{}
+	}
+	return nil
+}
+
+func (e *APIGatewayConfigEntry) validateMergedListeners() error {
+	listeners := make(map[int]APIGatewayListener)
```

Comment:
Good thought, I have compiled your "future notes" in [this doc](https://docs.google.com/document/d/1SmsakmGnrmd3uI3mthefHQkDrz6GvzMrQvaHoiKu15Y/edit?usp=sharing)

Reply:
This is fine for now, but in the Gateway API spec listeners don't have to be unique by port, but instead by the tuple of `{ hostname, port, protocol }` (ALPN isn't explicitly referenced in the spec yet, but I have heard some questions about how it might be used).

https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io%2fv1beta1.Gateway

Reply:
This is fine for now, but the Gateway API spec doesn't require unique ports per listener, only uniqueness on the tuple of `{ hostname, port, protocol }`, so the upstream flexibility should mostly exist for the switching Matt is describing as a future possibility.

https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io%2fv1beta1.Gateway